In [1]:
#این سلول استایل های لازم برای نمایش مناسب این دفترچه را ایجاد میکند. لطفا محنویات آن را تغییر ندهید
from IPython.display import HTML
titr_fa = "div.titr_fa { direction:rtl;padding: 0.5em;font-family:B Titr; font-size: 14pt;}"
norm_fa = "div.norm_fa { direction:rtl;text-align: justify;text-justify: inter-word;font-family:B Nazanin; font-size: 14pt;}"
warn_fa = "div.warn_fa { direction:rtl;background-color: #fcf2f2;border-color: #dFb5b4; border-left: 5px solid #dfb5b4; padding: 0.5em;text-align: justify;text-justify: inter-word;font-family:B Nazanin; font-size: 14pt;}"
style = "<style>" + titr_fa + norm_fa + warn_fa + "</style>"
HTML(style)

<div class="norm_fa">
    در این مطلب سراغ کتابخانه numpy میریم. numpy یک کتابخانه برای پیاده سازی آرایه های چند بعدی و انجام محاسبات سریع روی اونهاست. ابتدا بعد از فعال کردن محیط مجازی دلخواهمون یک دفترچه ژوپیتر ایجاد میکنیم (در مطلب قبل توضیح دادیم چجوری) . هدف اینه که این دفترچه تبدیل به مرجع آموزش و یادآوری numpy برای ما باشه. من برای تولید این دفترچه از <a href="https://scipy-lectures.org">این سایت</a>  خیلی استفاده کردم و برای کسب مهارت بیشتر میتونیم بهش رجوع کنیم. مستندات اصلی numpy برای رجوع رو در <a href="https://docs.scipy.org/doc/numpy/dev">این آدرس</a>  میتونیم پیدا کنیم. 
</div>

<div class="titr_fa">
    1- مقدمات
</div>

<div class="norm_fa">
 به جرات میشه گفت اگر پایتون جای خودش رو در جامعه آکادمیک باز کرده به خاطر این کتابخانه است. چرا؟ اول numpy رو با اسم مستعار np وارد میکنیم :
</div>

In [1]:
import numpy as np

 <div class="norm_fa">
    حالا آرایه های پایتون و numpy رو با هم مقایسه میکنیم. 
 </div>

In [2]:
%%time 
sq = [i**2 for i in range(1000000)]

Wall time: 387 ms


<div class="norm_fa">
    به دستوراتی که با %% در دفترچه شروع میشن میگن جادوی سلولی (cell magic) و فقط برای همون سلول یک کار جادویی انجام میدن. مثلا اینجا جادوی سلولی time زمان اجرای یک سلول رو اندازه میگیره. ایجاد و به توان 2 رسوندن یک لیست یک میلیون تایی در پایتون خالص در کامپیوتر من 387 میلی ثانیه زمان برد. حالا با numpy:
</div>

In [3]:
%%time
npsq = np.arange(1000000) ** 2

Wall time: 4.48 ms


<div class="norm_fa">
    یعنی حدود 70 برابر سریعتر. خود پایتون ذاتا زبان سریعی نیست ولی numpy با C نوشته شده که سرعت بالا و راحتی پایتون رو با هم ترکیب میکنه که برای انجام سریع کار تحیقاتی بسیار مناسبه. حالا میریم سراغ دستورات numpy. گفتیم که هدف numpy ایجاد یک کتابخانه برای تولید سریع و بهینه آرایه های چند بعدی و محاسبات سریع روی اونهاست. پس اولین مرحله تولید آرایه ست که با تابع array انجام میشه:
</div>

In [4]:
a = np.array([0,1,2,3])
a

array([0, 1, 2, 3])

<div class="norm_fa">
مشخصات این آرایه چیه؟
</div>

In [5]:
print(a.ndim)
print(a.shape)
print(len(a))

1
(4,)
4


<div class="norm_fa">
یعنی به ترتیب آرایه یک بعدیه، شکلش (,4) و طولش 4ه. Andrew Ng استاد معروف دانشگاه استانفورد، از پیشگامان Deep Learning و بنیانگذار Coursera، در دوره آموزشی Deep Learning Specialization ما رو از به کار بردن آرایه های یک بعدی، به خاطر ایرادات خاصی که در صورت بی دقتی میتونه ایجاد کنه، نهی میکنه. ما هم گوش میدیم و :</div>

In [6]:
b = np.array([[0,1,2,3]])
b

array([[0, 1, 2, 3]])

In [7]:
print(b.ndim)
print(b.shape)
print(len(b))

2
(1, 4)
1


<div class="norm_fa">
    حالا یک آرایه (ماتریس) دوبعدی داریم که یک سطر داره و 4 ستون. برای ایجاد آرایه دو بعدی باید یک آرایه از آرایه ها درست کنیم. هر سطر خودش یک آرایه از اعداده و آرایه این سطرها کنار هم ماتریس دوبعدی رو تشکیل میده :
</div>

In [8]:
c = np.array([[0,1,2,3],[4,5,6,7]])
c

array([[0, 1, 2, 3],
       [4, 5, 6, 7]])

In [9]:
print(c)

[[0 1 2 3]
 [4 5 6 7]]


<div class="norm_fa">
یعنی آرایه c خودش شامل 2 تا آرایه (سطر) است که هرکدوم 4 تا عنصر (ستون) دارند. دستور len طول اولین بعد (تعداد سطرها) را بر میگردونه:
</div>

In [10]:
len(c)

2

<div class="norm_fa">
معمولا ما آرایه ها رو اینجوری دستی وارد نمیکنیم. یا اونها رو میسازیم یا از فایل میخونیم.
دستور بعد تابع arangeه که یک توالی اعداد درست میکنه:
</div>

In [11]:
np.arange(10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

<div class="norm_fa">
مثلا این تابع 3 تا پارامتر هم میگیره که به ترتیب شروع توالی، پایان توالی و گام پرش هستند
    .مثلا اگر بخواهیم با شروع از 1، 2 تا 2تا بپریم و تا 10 این کار رو ادامه بدیم:
</div>

In [12]:
np.arange(1,10,2)

array([1, 3, 5, 7, 9])

<div class="norm_fa">
گام پرش میتونه عدد اعشاری باشه
</div>

In [13]:
np.arange(1,5, 0.4)

array([1. , 1.4, 1.8, 2.2, 2.6, 3. , 3.4, 3.8, 4.2, 4.6])

<div class="norm_fa">
برای تقسیم یک بازه با ابتدا و انتهای مشخص به تعداد معین از linspace استفاده میکنیم.مثلا از بازه 1 تا 10، 4 تا عدد با فاصله برابر میخواهیم:
</div>

In [14]:
np.linspace(1,10,4)

array([ 1.,  4.,  7., 10.])

<div class="norm_fa">
از بازه 0 تا 1 ، 11 عدد با فاصله برابر میخوایم:
</div>

In [15]:
np.linspace(0,1,11)

array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. ])

<div class="norm_fa">
فرق arange با linspace اینه که در arange گام پرش ثابته و در linspace تعداد نقاط.یک تابع logspace هم داریم که از بازه تعداد  نقاط ثابت با گامهای لگاریتمی استخراج میکنه، فقط باید لگاریتم مبنای 10 نقطه شروع و پایان رو بهش بدید:
</div>

In [16]:
np.logspace(np.log10(10),np.log10(100),15)

array([ 10.        ,  11.78768635,  13.89495494,  16.37893707,
        19.30697729,  22.75845926,  26.82695795,  31.6227766 ,
        37.2759372 ,  43.93970561,  51.79474679,  61.05402297,
        71.9685673 ,  84.83428982, 100.        ])

<div class="norm_fa">
که پانزده تا عدد بین 10 تا 100 بهمون میده که گامهای بینشون لگاریتمیه و به مرور فاصله اعداد با عدد قبلیشون بیشتر میشه.
</div>

<div class="norm_fa">
برای ساخت ماتریس که همه عناصرش 1 باشند از تابع ones و برای صفر از zeros استفاده میکنیم و ابعاد ماتریس رو به صورت یک tuple بهش میدیم:
</div>

In [17]:
np.ones((4, 4))

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [18]:
np.zeros((3,5))

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

<div class="norm_fa">
این دو تا تابع یک پارامتر دیگه به نام dtype هم میگیرن که نوع ریاضی اعضای ماتریس رو مشخص میکنه. مثلا اگر ماتریس ones از نوع مختلط بخواهیم:
</div>

In [19]:
np.ones((4, 4), dtype=np.complex)

array([[1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j],
       [1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j],
       [1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j],
       [1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j]])

<div class="norm_fa">
مقادیر معتبر برای dtype اینها هستند:
np.complex برای اعدا مختلط, 
np.int8, np.int16, np.int32, np.int64 برای اعداد صحیح علامتدار, 
np.float16, np.float32, np.float64 برای اعداد اعشاری, 
np.uint16, np.uint32, np.uint64 برای اعداد صحیح بدون علامت
</div>

<div class="norm_fa">
ماتریس واحد رو هم با تابع eye میسازیم:
</div>

In [20]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

<div class="titr_fa">
2- تولید اعداد تصادفی
</div>

<div class="norm_fa">
حالا سراغ تولید اعداد تصادفی میریم. دو تا توزیع مهم برای تولید اعداد تصادفی داریم : یکنواخت و گاوسی. اینا یعنی چی؟ از توزیع یکنواخت شروع میکنیم:
</div>

In [21]:
np.random.rand(1,20)

array([[0.3756344 , 0.90233306, 0.1375995 , 0.99084453, 0.97942387,
        0.55801556, 0.5568001 , 0.32760203, 0.80889496, 0.16096867,
        0.27003026, 0.09895287, 0.64828136, 0.23600891, 0.14034868,
        0.23665831, 0.7972643 , 0.98966386, 0.19744077, 0.40562954]])

<div class="norm_fa">
که برامون 20 تا عدد تصادفی بین 0 تا 1 تولید میکنه که شانس انتخاب شدن از همه جای این بازه با هم برابره. یعنی مثلا یک میلیون عدد تصادفی با این توزیع تولید کنید و تعداد اونهایی که بین 0 تا 0.1 هستند رو بشمارید احتمالا نزدیک تعداد اونهایی که بین 0.9 تا 1 هستند.(یا هر بازه دیگر با طول 0.1). اگر بخواهیم به جای صفر و یک، بین a,b برامون اعداد تصادفی تولید بشه از این فرمول استفاده میکنیم:
</div>

In [22]:
#(b-a) * np.random.rand(ابعاد ماتریس دلخواه) + a

<div class="norm_fa">
مثلا 20 عدد بین 2و5
</div>

In [23]:
(5-2)*np.random.rand(1,20) + 2

array([[4.35316675, 2.18119736, 3.43178322, 4.76958118, 2.37552475,
        4.23445577, 2.03242133, 4.13035807, 4.63918657, 4.38730227,
        4.72660014, 3.57864619, 3.31436614, 4.30411698, 3.84090825,
        2.98123599, 4.99701824, 4.85172761, 4.46806026, 4.65936171]])

<div class="norm_fa">
اما اعداد تصادفی گاوسی یا نرمال یا استاندارد با randn تولید میشن و توزیعشون فرق داره.
</div>

In [24]:
np.random.randn(1,20)

array([[-0.20268403,  1.24981011, -0.32934498, -0.95969644,  0.11731372,
         0.73241878,  0.64427743,  1.09750409, -0.66356479,  1.01457575,
         1.07618613,  0.75146773, -0.52147204, -1.03646218, -0.05561395,
        -1.61663065,  0.74230641, -0.70751792, -0.58220838, -0.82017621]])

<div class="norm_fa">
ساده بخوام یگم اگر تعداد زیادی از اعداد تصادفی با این توزیع درست کنید یه احتمال 65 درصد اعداد بین 1و 1- و به احتمال 95 درصد بین 2و 2- و به احتمال 99.7 درصد بین 3 و 3- هستند. یا این جوری بگم که تراکمشون حول صفر زیادتره و هرچی از صفر دور بشیم تراکم اعداد هم کمتر میشه. یعنی از یک توزیع گوسی N(0,1) با میانگین صفر و واریانس 1 تولید میشن. اگر بخواهیم میانگین توزیع رو از صفر به mu و واریانس رو از 1 به sigma^2 تغییر بدیم از فرمول زیر استفاده میکنیم: 
</div>

In [25]:
#sigma * np.random.randn(ابعاد ماتریس دلخواه) + mu

<div class="norm_fa">
برای تولید اعداد تصادفی صحیح از randint استفاده میکنیم. برای تولید یک آرایه 2 در 20 از اعداد تصادفی صحیح بین 0 تا 10 اینجوری مینویسیم:
</div>

In [26]:
np.random.randint(0,10, (2,20))

array([[9, 8, 6, 7, 7, 9, 2, 9, 6, 1, 7, 2, 8, 5, 5, 0, 1, 5, 5, 5],
       [4, 9, 9, 9, 5, 8, 9, 8, 4, 4, 4, 7, 4, 4, 9, 7, 5, 7, 8, 7]])

<div class="norm_fa">
لازم به توضیح که این اعداد شبه تصادفین و با الگوریتم Mersenne Twister ایجاد میشن. برای همین میتونیم تولید شدنشون رو با تابع seed کنترل کنیم. تابع seed یک عدد میگیره و بعد از اجرا شدنش ترتیب اعداد تولید شده یکسان ولی شبه تصادفی خواهد بود. مثلا:
</div>

In [27]:
np.random.seed(201)
print(np.random.rand(1,10))
print(np.random.rand(1,10))

[[0.1139046  0.74788082 0.6110594  0.8096918  0.71285749 0.14003681
  0.26751188 0.14760445 0.22806792 0.23556173]]
[[0.69002382 0.14834782 0.44560583 0.70084254 0.09663865 0.88263259
  0.86983229 0.84459421 0.27370208 0.78567954]]


<div class="norm_fa">
ما اینجا دو بار تابع rand رو بعد از تنظیم seed به 201 فراخوانی کردیم و این اعداد تولید شدند حالا اگه دو باره seed رو تنظیم کنیم:
</div>

In [28]:
np.random.seed(201)
print(np.random.rand(1,10))
print(np.random.rand(1,10))

[[0.1139046  0.74788082 0.6110594  0.8096918  0.71285749 0.14003681
  0.26751188 0.14760445 0.22806792 0.23556173]]
[[0.69002382 0.14834782 0.44560583 0.70084254 0.09663865 0.88263259
  0.86983229 0.84459421 0.27370208 0.78567954]]


<div class="norm_fa">
میبینیم که دو باره این اعداد شبه تصادفی با همون ترتیب تکرار شدند. این کارکرد برای ایراد یابی برنامه هایی که از اعداد تصادفی استفاده میکنند لازمه. اگر بخواهیم دیگه خیلی (شبه) تصادفی باشیم seed رو بدون پارامتر صدا کنیم. اما یک کتابخونه دیگه به نام rdrand هست که از مکانیزم تولید اعداد تصادفی واقعی داخل CPU شما استفاده میکنه و کیفیت تصادفی بودن اعداد تولید شده با اون خیلی بالاتره.
</div>

<div class="titr_fa">
3- کار با اندیس ها و برش ها
</div>

<div class="norm_fa">
برای دسترسی به عناصر مختلف در ماتریس، اندیس اون عنصر را بین [] وارد میکنیم. اندیس ها از صفر شروع میشن. همچنین اعداد منفی برای دسترسی از انتها استفاده میشن. اندیس 1- معادل آخری، 2- معادل یکی مونده به آخر و ... هستند. یک برش از آرابه خودش یک آرایست و همه کارهای آرایه رو میشه روش انجام داد.
</div>

In [29]:
d = np.arange(1, 10)
d

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [30]:
d[0], d[5], d[8]

(1, 6, 9)

In [31]:
d[-1], d[-4], d[-9]

(9, 6, 1)

In [32]:
e = np.array([[1,2,3], [4,5,6], [7,8,9]])
e

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [33]:
e[0,0], e[1, 1], e[2,1]

(1, 5, 8)

<div class="norm_fa">
برای دسترسی به یک برش از آرایه، شروع و پایان اون قطاع رو با یک : ازهم جدا میکنیم. مثلا برش 6 تا 7 از ماتریس d میشه:
</div>

In [34]:
d[6:8]

array([7, 8])

<div class="norm_fa">
دقت میکنیم که خود عددی که در قسمت پایان برش نوشته بشه در برش نمیاد و تا قبل از اون در برش منعکس میشن. در مثال بالا هم خود اندیس شماره 8 در قطاع وارد نمیشه و اندیس های 6 و 7 اومدن. میتونیم یک گام پرش هم به برش اضافه کنیم که البته اختیاریه:
</div>

In [35]:
d[3:9:2]

array([4, 6, 8])

<div class="norm_fa">
و اینجوری میخونیم: از ماتریس d اندیس های بین 3 تا 9 رو با پرش 2 تا 2 تا به من بده (حواسم هم هست که خود اندیس 9 در برش نمیاد). میشه جای هرکدوم از بخش های برش رو خالی گذاشت. اگر جای شروع چیزی ننویسیم یعنی از ابتدا شروع کن. اگر جای پایان برش چیزی ننویسیم یعنی تا انتها برو. مثلا میگیم تا اندیس شماره 4 (یعنی اندیس های 0و 1و 2و 3) رو بهم بده:
</div>

In [36]:
d[:4]

array([1, 2, 3, 4])

<div class="norm_fa">
یا از اندیس 6 به بعد (معادل اندیس های 6و 7و 8 در ماتریس d)
 رو بهم بده:
</div>


In [37]:
d[6:]

array([7, 8, 9])

<div class="norm_fa">
از اعداد منفی هم در هر کدوم از ابتدا و انتها و گام پرش میتونبم استفاده کنیم ولی در گام پرش جای ابتدا و انتها رو عوض میکنه. مثلا میخوایم بگیم یک برش از ماتریس d از انتها به ابتدا، به صورت برعکس به من بده:
</div>

In [38]:
d[::-1]

array([9, 8, 7, 6, 5, 4, 3, 2, 1])

In [39]:
d[::3] # (گام پرش 3) یه صورت دو تا در میون

array([1, 4, 7])

In [40]:
d[4::-1] # از اندیس 4 تا ابتدا به صورت برعکس

array([5, 4, 3, 2, 1])

In [41]:
d[1:-1] # عنصر اول و آخر رو حذف کن

array([2, 3, 4, 5, 6, 7, 8])

In [42]:
d[-3:] #از 2 تا مونده به آخر تا آخر رو به من بده

array([7, 8, 9])

<div class="norm_fa">
برای ماتریس های دو بعدی هم همینطوره و برای هرکدوم از ابعاد میتونید یک برش بنویسید. این عکس که از <a href="https://scipy-lectures.org">این سایت</a>  برداشتم خیلی خوب این مفهوم رو توصیف میکنه:
</div>

![توصیف تصویری برش ها در ماتریس با ابعاد بالاتر](./notebook_imgs/np_indexing.png)

<div class="norm_fa">
مخصوصا میخوام روی آبیه و قرمزه تاکید کنم. آبیه میگه تمام سطرها، از ستون با اندیس 2 ماتریس a رو به من بده. قرمزه میگه از سطر با اندیس 0 ، ستونهای بین اندیس های 4و 5 رو به من بده. 
</div>

<div class="norm_fa">
از برشها و اندیس ها برای مقداردهی عناصر آرایه هم میشه استفاده کرد. مثلا میخوایم بگیم همه عناصر بعد از اندیس 5 ماتریس d رو به 10 تغییر بده:
</div>

In [43]:
d[5:] = 10
d

array([ 1,  2,  3,  4,  5, 10, 10, 10, 10])

<div class="norm_fa">
یا اینکه ترتیب عناصر  تا اندیس 5 ماتریس d رو برعکس کن:
</div>

In [44]:
d[:5] = d[4::-1]
d

array([ 5,  4,  3,  2,  1, 10, 10, 10, 10])

<div class="norm_fa">
میشه از یک ماتریس هم برای اندیس گذاری و برش یک ماتریس دیگه استفاده کرد. مثلا میخواهیم از ماتریس d فقط اندیسهای 1,3,5,7 رو استخراج کنیم:
</div>

In [49]:
d[[1,3,5,7]]

array([ 4,  2, 10, 10])

<div class="norm_fa">
برای ماتریس های دو (یاچند) بعدی هم ممکنه: 
</div>

In [50]:
f = np.array([[1,2,3,4,5],[6,7,8,9,10],[11,12,13,14,15]])
f

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10],
       [11, 12, 13, 14, 15]])

<div class="norm_fa">
میخواهیم عناصر اندیسهای (0,1) و (1,2) و (2,4) رو استخراج کنیم:
</div>

In [54]:
f[[0,1,2],[1,2,4]]

array([ 2,  8, 15])

<div class="norm_fa">
یا بهشون مقدار بدیم:
</div>

In [55]:
f[(0,1,2),(1,2,4)] = -1
f

array([[ 1, -1,  3,  4,  5],
       [ 6,  7, -1,  9, 10],
       [11, 12, 13, 14, -1]])

<div class="norm_fa">
حالا میریم سراغ اندیس گذاری منطقی.  ماتریس ها رو میشه در یک گزاره منطقی به کار برد که خروجی اون یک ماتریس با ابعاد ماتریس اصلیه و به جای عناصری که در شرط صدق میکن True و به جای بقیه False اومده. حالا برای آزمایش اول یک ماتریس تصادفی درست میکنیم:
</div>

In [56]:
np.random.seed(3)
g = np.random.randint(0, 20, (1,20))
g

array([[10,  3,  8,  0, 19, 10, 11,  9, 10,  6,  0, 12,  7, 14, 17,  2,
         2,  1, 19,  5]])

<div class="norm_fa">
 و میخوایم ببینیم کدام یک از عناصرش بر 3 بخشپذیره:
</div>

In [57]:
g%3==0

array([[False,  True, False,  True, False, False, False,  True, False,
         True,  True,  True, False, False, False, False, False, False,
        False, False]])

<div class="norm_fa">
میبینیم که در این ماتریس معادل اندیس هایی در ماتریس g که اعداد اون اندیس بر 3 بخشپذیرند True درج شده. حالا اگر از این ماتریس برای برش ماتریس g استفاده کنیم:
</div>

In [58]:
g[g%3==0]

array([ 3,  0,  9,  6,  0, 12])

<div class="norm_fa">
    میبینیم فقط عناصری از g اومدن که در ماتریس دوم معادل اندیسشون True درج شده (بر 3 بخشپذیرن). اندیس گذاری منطقی رو مبشه با عملگر های دیگه هم به کار برد:
</div>

In [59]:
g[g>10]

array([19, 11, 12, 14, 17, 19])

In [60]:
g[g<=5]

array([3, 0, 0, 2, 2, 1, 5])

<div class="norm_fa">
یا برای مقدار دهی ازشون استفاده کرد.مثلا میخواهیم اون عناصری که بر 2 بخشپذیر نیستند رو صفر کنیم:
</div>

In [61]:
g[g%2!=0] = 0
g

array([[10,  0,  8,  0,  0, 10,  0,  0, 10,  6,  0, 12,  0, 14,  0,  2,
         2,  0,  0,  0]])

<div class="titr_fa">
4- عملیات اعداد با آرایه ها
</div>

In [62]:
h = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9])
h

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [63]:
h + 1

array([ 2,  3,  4,  5,  6,  7,  8,  9, 10])

In [64]:
h-3

array([-2, -1,  0,  1,  2,  3,  4,  5,  6])

In [65]:
2*h

array([ 2,  4,  6,  8, 10, 12, 14, 16, 18])

In [66]:
h/3

array([0.33333333, 0.66666667, 1.        , 1.33333333, 1.66666667,
       2.        , 2.33333333, 2.66666667, 3.        ])

<div class="norm_fa">
عملگر به توان رسوندن در پایتون **ه:
</div>

In [67]:
2**h

array([  2,   4,   8,  16,  32,  64, 128, 256, 512], dtype=int32)

In [68]:
h**2

array([ 1,  4,  9, 16, 25, 36, 49, 64, 81], dtype=int32)

<div class="norm_fa">
توان میتونه عدد صحیح نباشه. مثلا اگه بخواهیم جذر عناصر h رو بگیریم:
</div>

In [69]:
h**0.5

array([1.        , 1.41421356, 1.73205081, 2.        , 2.23606798,
       2.44948974, 2.64575131, 2.82842712, 3.        ])

<div class="titr_fa">
5 - توابع پایه ریاضی:
</div>

In [70]:
i = np.sqrt(h) # جذر عناصر آرایه
i

array([1.        , 1.41421356, 1.73205081, 2.        , 2.23606798,
       2.44948974, 2.64575131, 2.82842712, 3.        ])

In [71]:
np.abs(np.array([-1, -3, 5, 6, -9, -3, 2])) #قدر مطلق 

array([1, 3, 5, 6, 9, 3, 2])

<div class="norm_fa">
 j در نامپای برای ساختن اعداد مختلط به کار میره و بهتره برای اسم ماتریس ازش استفاده نکنیم:
</div>

In [72]:
k = np.array([1+1j, 2+3j, 6-7j, -2j, 1])
k

array([ 1.+1.j,  2.+3.j,  6.-7.j, -0.-2.j,  1.+0.j])

In [73]:
l = np.pi * h /2 # عدد pi 
l

array([ 1.57079633,  3.14159265,  4.71238898,  6.28318531,  7.85398163,
        9.42477796, 10.99557429, 12.56637061, 14.13716694])

In [74]:
np.sin(l) # سینوس هر عنصر، این تابع فرض میکند که عناصر آرایه بر حسب رادیان هستند، نه درجه

array([ 1.0000000e+00,  1.2246468e-16, -1.0000000e+00, -2.4492936e-16,
        1.0000000e+00,  3.6739404e-16, -1.0000000e+00, -4.8985872e-16,
        1.0000000e+00])

<div class="norm_fa">
ما اون e رو به صورت "ضربدر 10 به توان" میخونیم. مثلا 1e-16 میشه 1 ضربدر 10 به توان منفی 16، که البته عدد خیلی کوچیکیه و نزدیک به صفر.
</div>

In [75]:
np.cos(l) # تابع کسینوس، توضیحات مانند سینوس

array([ 6.1232340e-17, -1.0000000e+00, -1.8369702e-16,  1.0000000e+00,
        3.0616170e-16, -1.0000000e+00, -4.2862638e-16,  1.0000000e+00,
        5.5109106e-16])

In [76]:
np.tan(l) # تانژانت

array([ 1.63312394e+16, -1.22464680e-16,  5.44374645e+15, -2.44929360e-16,
        3.26624787e+15, -3.67394040e-16,  2.33303419e+15, -4.89858720e-16,
        1.81458215e+15])

In [77]:
m = np.arctan(h) # آرکتانژانت، خروجی برحسب رادیان است
m

array([0.78539816, 1.10714872, 1.24904577, 1.32581766, 1.37340077,
       1.40564765, 1.42889927, 1.44644133, 1.46013911])

<div class="norm_fa">
اگر میخواهیم معادل درجه بک زاویه برحسب رادیان را ببینیم از تابع rad2deg استفاده میکنیم:
</div>

In [78]:
np.rad2deg(m)

array([45.        , 63.43494882, 71.56505118, 75.96375653, 78.69006753,
       80.53767779, 81.86989765, 82.87498365, 83.65980825])

In [79]:
n = np.exp(h) # عدد نپر به توان اعضای آرایه
n

array([2.71828183e+00, 7.38905610e+00, 2.00855369e+01, 5.45981500e+01,
       1.48413159e+02, 4.03428793e+02, 1.09663316e+03, 2.98095799e+03,
       8.10308393e+03])

In [80]:
np.log10(h) # لگاریتم در مبنای 10 عناصر آرایه

array([0.        , 0.30103   , 0.47712125, 0.60205999, 0.69897   ,
       0.77815125, 0.84509804, 0.90308999, 0.95424251])

<div class="norm_fa">
لگاریتم پایه طبیعی در numpy با تابع log حساب میشه. من تو دبیرستان با اسم ln میشناختم. معکوس تابع exp:
</div>

In [81]:
np.log(n) # 

array([1., 2., 3., 4., 5., 6., 7., 8., 9.])

<div class="norm_fa">
روی اعداد مختلط هم میتونیم محاسبه کنیم:
</div>

In [82]:
np.abs(k) # اندازه

array([1.41421356, 3.60555128, 9.21954446, 2.        , 1.        ])

In [83]:
np.angle(k) # زاویه برحسب رادیان

array([ 0.78539816,  0.98279372, -0.86217005, -1.57079633,  0.        ])

In [84]:
np.real(k) # بخش حقیقی

array([ 1.,  2.,  6., -0.,  1.])

In [85]:
np.imag(k) # بخش موهومی

array([ 1.,  3., -7., -2.,  0.])

In [86]:
np.conj(k) # مزدوج مختلط

array([ 1.-1.j,  2.-3.j,  6.+7.j, -0.+2.j,  1.-0.j])

<div class="titr_fa">
6- عملیات آرایه با آرایه (ماتریس با ماتریس)
</div>

<div class="norm_fa">
عملیات عنصر به عنصر
</div>

In [87]:
o = np.ones((4,4))
o

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [88]:
o + o # جمع عناصر در اندیس های متناظر

array([[2., 2., 2., 2.],
       [2., 2., 2., 2.],
       [2., 2., 2., 2.],
       [2., 2., 2., 2.]])

In [89]:
np.add(o,o) # معادل o+o

array([[2., 2., 2., 2.],
       [2., 2., 2., 2.],
       [2., 2., 2., 2.],
       [2., 2., 2., 2.]])

In [90]:
o / (o+o)# تقسیم عناصر در اندیس های متناظر

array([[0.5, 0.5, 0.5, 0.5],
       [0.5, 0.5, 0.5, 0.5],
       [0.5, 0.5, 0.5, 0.5],
       [0.5, 0.5, 0.5, 0.5]])

In [91]:
np.divide(o, o+o) # معادل تقسیم عنصر به عنصر

array([[0.5, 0.5, 0.5, 0.5],
       [0.5, 0.5, 0.5, 0.5],
       [0.5, 0.5, 0.5, 0.5],
       [0.5, 0.5, 0.5, 0.5]])

In [92]:
o - np.eye(4) # eye ماتریس واحد

array([[0., 1., 1., 1.],
       [1., 0., 1., 1.],
       [1., 1., 0., 1.],
       [1., 1., 1., 0.]])

In [93]:
np.subtract(o, np.eye(4))

array([[0., 1., 1., 1.],
       [1., 0., 1., 1.],
       [1., 1., 0., 1.],
       [1., 1., 1., 0.]])

In [94]:
o * o # ضرب عنصر در عنصرهای متناظر

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [95]:
np.multiply(o,o) #معادل o*o

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [96]:
np.dot(o, o) # ضرب ماتریس در ماتریس

array([[4., 4., 4., 4.],
       [4., 4., 4., 4.],
       [4., 4., 4., 4.],
       [4., 4., 4., 4.]])

<div class="norm_fa">
در تمامی این عملیاتها باید ابعاد ماتریسها با هم همخوانی داشته باشد در غیر اینصورت:
</div>

In [97]:
np.ones((4,4)) + np.ones((3,3))

ValueError: operands could not be broadcast together with shapes (4,4) (3,3) 

<div class="norm_fa">
میگه آرایه ها رو باهم نمیتونه پخش (broadcast) کنه. راجع به پخش در ادامه صحبت خواهیم کرد. برای ضرب دو ماتریس ابعاد داخلیشون باید با هم برابر باشه و گرنه:
</div>

In [98]:
np.dot(np.ones((4,4)) , np.ones((3,4)))

ValueError: shapes (4,4) and (3,4) not aligned: 4 (dim 1) != 3 (dim 0)

<div class="norm_fa">
روی ماتریس ها عملیات منطقی رو هم میتونیم انجام بدیم. طبق یک قراداد قدیمی که از زبان C به ارث رسیده، عدد 0 معادل غلط (False) و 1 (یا هر چیز دیگه به جز 0) معادل صحیح (True) در نظر گرفته میشه:
</div>

In [99]:
o == np.eye(4) # برابری عنصر به عنصر

array([[ True, False, False, False],
       [False,  True, False, False],
       [False, False,  True, False],
       [False, False, False,  True]])

In [100]:
p = np.array([0, 0, 1, 1])
q = np.array([0, 1, 0, 1])

In [101]:
np.logical_and(p,q)

array([False, False, False,  True])

In [102]:
np.logical_or(p,q)

array([False,  True,  True,  True])

In [103]:
np.logical_xor(p,q)

array([False,  True,  True, False])

In [104]:
np.logical_not(p)

array([ True,  True, False, False])

In [105]:
R = np.array([[0,1,2,3,4], [5,6,7,8,9], [10,11,12,13,14], [15,16, 17,18,19]])
R

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19]])

<div class="norm_fa">
در عملیات آرایه با آرایه numpy یک مفهومی داره به اسم پخش (broadcast). مثلا ما میدونیم در جمع دو ماتریس، هر دو ماتریس جمع شونده باید سطر و ستونهای یکسان داشته باشند. اما مثلا ما یک عدد صحیح (که بعدش 1ه) رو با یک ماتریس دو بعدی جمع کردیم. اینجا  numpy عملیات پخش رو انجام میده. مثلا اگر یک ماتریس (3,3) رو بخواهیم با یک ماتریس (1,3) جمع کنیم، numpy سطر اول ماتریس دوم رو 3 بار زیر هم کپی میکنه تا یک ماتریس (3,3) ایجاد کنه و هر دو طرف عملگر جمع هم اندازه بشن و بعد اونها رو با هم جمع میکنه. در عملیات جزء به جزء هر وقت یکی از ابعاد ماتریس ها با هم بخونه، پخش انجام میشه. ماتریس 1 در 1 (یک عدد تنها) به هر ابعادی میتونه پخش بشه.
</div>

<div class="norm_fa">
و بالاخره ترانهاده یک ماتریس:
</div>

In [106]:
R.T

array([[ 0,  5, 10, 15],
       [ 1,  6, 11, 16],
       [ 2,  7, 12, 17],
       [ 3,  8, 13, 18],
       [ 4,  9, 14, 19]])

<div class="titr_fa">
7- کاهش (Reduction):
</div>

<div class="norm_fa">
کاهش ها عملیاتی هستند که یک ورودی آرایه بزرگ میگیرن و روی اون محاسباتی انجام میدن و یک آرایه کوچکتر و یا یک عدد خروجی میدن. مثلا تابع مجموع:
</div>

In [107]:
np.sum(R)

190

<div class="norm_fa">
تقریبا در همه این توابع میتونیم بجای کاهش کل ماتریس، روی سطرها یا ستونها (و اگر ابعاد بیشتری دارند، روی اونها هم) به صورت مجزا کاهش انجام بدیم. پارامتر axis  محور مورد نظر برای کاهش رو مشخص میکنه. محور 0، سطرهای یک ماتریس رو کاهش میده. محور 1 ستونها و به همین ترتیب:
</div>

In [108]:
np.sum(R, axis = 0)

array([30, 34, 38, 42, 46])

<div class="norm_fa">
دستور بالا سطرها رو کاهش میده و میکنه یک سطر. حالا اون یک سطر چجوری ساخته میشه؟ : 
</div>

<div class="norm_fa">
اگر تصویری میخواید فکر کنید تمام سطرها رو در نظر بگیرید که به هم نزدیک میشن و فشرده میشن و آخرش میمونه 1 سطر که چند تا ستون داره. اعضای اون یک سطر چی هستن؟ اولین ستون مجموع تمام عناصریه که قبلا در ستون اول و در سطرهای مختلف بودن.
</div>

<div class="norm_fa">
اگر برنامه نویس هستید R[i,j] رو در نظر بگیرید که i اندیس سطرهاست و اول اومده و j اندیس ستونهاست که دوم اومده .اگر ابعاد ماتریس بیشتر باشه کاهش روی اونها رو هم میتونیم در نظر بگیریم. کاهش روی محور صفر (axis = 0)  یعنی یک حلقه روی تمامی مقادیر ممکن برای i در حالی که j (و بقیه اندیسها) ثابت هستند و جمع همه اونها.
</div>

<div class="norm_fa">
اگر هم حوصله ندارید، حفظ کنید که محور 0 از بالا به پایینه و محور 1 از چپ به راست. محور 2 در عمقه و برای ابعاد بالاتر رو هم یه کاریش بکنید.
</div>

In [109]:
np.sum(R, axis = 1)

array([10, 35, 60, 85])

<div class="norm_fa">
axis = 1 ستونها رو کاهش میده و میکنه یک ستون.
</div>

<div class="norm_fa">
چون R یک آرایه numpyه، معمولا میشه این دستورات رو به شکل دیگه ای هم نوشت:
</div>

In [110]:
R.sum()

190

In [111]:
R.sum(axis = 0)

array([30, 34, 38, 42, 46])

<div class="norm_fa">
حالا میریم سراغ کاهشهای دیگه:
</div>

In [112]:
R.min()#کمینه کل ماتریس

0

In [113]:
R.min(axis = 1) # کمینه محور 1. هر ستون کمینه تمام عناصریه که در اون ستون ولی در سطرهای مختلف بودن

array([ 0,  5, 10, 15])

In [114]:
R.max(axis = 0) # بیشنه محور 0. عمودی

array([15, 16, 17, 18, 19])

In [115]:
R.argmax(axis = 1) # اندیس عنصر بیشینه. تمام عناصر بیشینه در ستون آخر با اندیس 4 هستند

array([4, 4, 4, 4], dtype=int64)

In [116]:
R.argmin(axis = 0) # اندیس عنصر کمینه در هر محور 0. همگی در سطر اول هستند

array([0, 0, 0, 0, 0], dtype=int64)

In [117]:
R.mean() # میانگین کل ماتریس

9.5

<div class="norm_fa">
قبلا گفتیم برشهای یک ماتریس خودشون یک ماتریس هستن:
</div>

In [118]:
R[:,:2].mean() #میانگین کل برش تمام سطرهای دو ستون اول

8.0

<div class="norm_fa">
و میشه محورها رو هم اعمال کرد تا قشنگ سردرد بگیریم:
</div>

In [119]:
R[:-1,:].mean(axis = 1) #میانگین روی محور 1 برش همه سطرها به جز آخری و همه ستونها

array([ 2.,  7., 12.])

In [120]:
np.median(R, axis = -1) # میانه آخرین محور. در اینجا محور 1

array([ 2.,  7., 12., 17.])

<div class="norm_fa">
قبلا گفتیم که عدد صفر ومیتونیم جای False و غیرصفر رو جای True استفاده کنیم. همچینین گفتیم که در اندیس گذاری منطقی خروجی یک ماتریس با ابعاد ماتریس اصلیه که به جای عناصری که در شرط صدق میکن True و به جای بقیه False اومده. حالا اگر از ترکیب تابع count_nonzero و اندیس های منطقی استفاده کنیم میتونیم تعداد عناصری از آرایه که در اون شرط صدق کردن رو بشمریم:
</div>

In [121]:
np.count_nonzero(R < 10)

10

In [122]:
R > 7

array([[False, False, False, False, False],
       [False, False, False,  True,  True],
       [ True,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True]])

In [149]:
np.count_nonzero(R > 7, axis = 1)

array([0, 2, 5, 5], dtype=int64)

<div class="titr_fa">
8- تغییر ابعاد آرایه ها
</div>

<div class="norm_fa">
اگر بخواهیم یک آرایه چند بعدی رو تخت کنیم، یعنی به یک ماتریس یک بعدی تبدیل کنیم از تابع flatten استفاده میکنیم:
</div>

In [129]:
S = R.flatten() # سطرها را پشت سر هم ردیف کرده و یک ماتریس یک بعدی میسازد
S

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19])

<div class="norm_fa">
اگر آرایه دارای ابعاد بیشتری باشد، ابعاد از انتها به ابتدا تخت میشوند. معکوس این تابع reshapeه :
</div>

In [130]:
S.reshape((2,10))

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14, 15, 16, 17, 18, 19]])

<div class="norm_fa">
اگر تعداد یکی از ابعاد رو حوصله ندارید محاسبه کنید، میتونید جاش 1- بگذارید تا خود numpy براتون مقدارش رو استنتاج کنه. مثلا میدونیم آرایه رو 5 سطری میخواهیم:
</div>

In [131]:
S.reshape((5,-1))

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15],
       [16, 17, 18, 19]])

<div class="norm_fa">
اگر بخواهیم به یک ماتریس یک بعد اضافه کنیم میتونیم از np.newaxis به این شکل استفاده کنیم:
</div>

In [132]:
Q = S[:, np.newaxis]
print(S.shape)
print(Q.shape)

(20,)
(20, 1)


In [133]:
Q[:, np.newaxis].shape

(20, 1, 1)

<div class="norm_fa">
برای چسبوندن دو ماتریس به هم از توابع vstackو hstack میتونیم استفاده کنیم. این توابع همه ماتریس هایی روی که میخواهیم به هم بچسبونیم رو از ما میگیرن و به ترتیب به صورت عمودی و افقی به هم میچسبونند :
</div>

In [143]:
print('a = ')
print(a)
print('c = ')
print(c)

a = 
[0 1 2 3]
c = 
[[0 1 2 3]
 [4 5 6 7]]


In [141]:
np.vstack((a, c)) # ماتریس ها به ترتیبی که ذکر شدن زیرهمدیگه آورده میشن و به هم میچسبن

array([[0, 1, 2, 3],
       [0, 1, 2, 3],
       [4, 5, 6, 7]])

In [144]:
np.vstack((c, a))

array([[0, 1, 2, 3],
       [4, 5, 6, 7],
       [0, 1, 2, 3]])

In [145]:
np.hstack((S,a)) # کنار همدیگه میچسبن

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19,  0,  1,  2,  3])

In [146]:
np.hstack((c,c))

array([[0, 1, 2, 3, 0, 1, 2, 3],
       [4, 5, 6, 7, 4, 5, 6, 7]])

In [217]:
np.resize(S,30) #تغییر اندازه ماتریس . عناصری که از قبل موجود نبودند با عناصر قبلی پر میشن

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9])

In [220]:
np.resize(c, (8,8))

array([[0, 1, 2, 3, 4, 5, 6, 7],
       [0, 1, 2, 3, 4, 5, 6, 7],
       [0, 1, 2, 3, 4, 5, 6, 7],
       [0, 1, 2, 3, 4, 5, 6, 7],
       [0, 1, 2, 3, 4, 5, 6, 7],
       [0, 1, 2, 3, 4, 5, 6, 7],
       [0, 1, 2, 3, 4, 5, 6, 7],
       [0, 1, 2, 3, 4, 5, 6, 7]])

<div class="titr_fa">
9 - سایر توابع مهم
</div>

In [165]:
np.random.seed(7)
t = np.random.randint(0, 20, (5,5))
t

array([[15,  4,  3, 19,  7],
       [14,  8, 14, 10,  8],
       [ 7,  6,  4, 16,  7],
       [12,  0, 11,  6, 19],
       [12,  5, 16,  0,  2]])

In [166]:
t.sort(axis = 0) # مرتب سازی هر ستون به صورت مجزا. 
t

array([[ 7,  0,  3,  0,  2],
       [12,  4,  4,  6,  7],
       [12,  5, 11, 10,  7],
       [14,  6, 14, 16,  8],
       [15,  8, 16, 19, 19]])

In [167]:
t.sort(axis = 1) # مرتب سازی هر سطر به صورت مجزا
t

array([[ 0,  0,  2,  3,  7],
       [ 4,  4,  6,  7, 12],
       [ 5,  7, 10, 11, 12],
       [ 6,  8, 14, 14, 16],
       [ 8, 15, 16, 19, 19]])

In [176]:
np.around([0.1, 0.7, -0.1, -0.7, 1.7, -1.7, 2.1, -2.1]) #  روند کردن به سمت نزدیک ترین عدد صحیح

array([ 0.,  1., -0., -1.,  2., -2.,  2., -2.])

In [177]:
np.ceil([0.1, 0.7, -0.1, -0.7, 1.7, -1.7, 2.1, -2.1]) #  روند کردن به سمت مثبت بینهایت(سقف)

array([ 1.,  1., -0., -0.,  2., -1.,  3., -2.])

In [178]:
np.floor([0.1, 0.7, -0.1, -0.7, 1.7, -1.7, 2.1, -2.1]) #  روند کردن به سمت منفی بینهایت(کف)

array([ 0.,  0., -1., -1.,  1., -2.,  2., -3.])

In [179]:
np.fix([0.1, 0.7, -0.1, -0.7, 1.7, -1.7, 2.1, -2.1]) # (جزء صحیح) روند کردن به سمت صفر

array([ 0.,  0., -0., -0.,  1., -1.,  2., -2.])

In [183]:
print(S)
print(np.cumsum(S)) #  هر عنصر مجموع تمام عناصر قبلی بعلاوه همان عنصر

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]
[  0   1   3   6  10  15  21  28  36  45  55  66  78  91 105 120 136 153
 171 190]


In [184]:
np.cumsum(t, axis = 0)

array([[ 0,  0,  2,  3,  7],
       [ 4,  4,  8, 10, 19],
       [ 9, 11, 18, 21, 31],
       [15, 19, 32, 35, 47],
       [23, 34, 48, 54, 66]], dtype=int32)

In [191]:
np.diff([1,2,4,7,0,5,-2]) # تفاضل هر عنصر با عنصر قبلی. عنصر اول در نظر گرفته نمیشود

array([ 1,  2,  3, -7,  5, -7])

In [225]:
u = np.fft.fft(np.cos(2*np.pi*np.linspace(0,2*np.pi,16))) # fft(cos(2pi*x)), x = [0, pi/8, pi/4, ..., 2pi]
u

array([ 0.52393458+0.j        ,  0.51866465+0.13808341j,
        0.50087511+0.29015025j,  0.46277537+0.47705368j,
        0.38142755+0.74273557j,  0.17307063+1.22188196j,
       -0.75695611+2.76524937j,  5.21756757-4.86596333j,
        2.48121588+0.j        ,  5.21756757+4.86596333j,
       -0.75695611-2.76524937j,  0.17307063-1.22188196j,
        0.38142755-0.74273557j,  0.46277537-0.47705368j,
        0.50087511-0.29015025j,  0.51866465-0.13808341j])

<div class="norm_fa">
    برای دیدن fft خانوادش به <a href="https://docs.scipy.org/doc/numpy/reference/routines.fft.html">این لینک</a> رجوع کنید. 
</div>

In [206]:
np.histogram([1,2,5,1,3,4,5,7,6,8,3,1,2,2,5,3,3], bins = 5) # هیستوگرام 5 با 5 بین

(array([6, 4, 4, 1, 2], dtype=int64), array([1. , 2.4, 3.8, 5.2, 6.6, 8. ]))

<div class="norm_fa">
هیستوگرام دو ماتریس بر میگردونه. اولیش تعداد عناصر در هر بینه. ماتریس دوم ابتدا و انتهای هر بین رو مشخص میکنه. در مثال بالا 6 عدد بین 1 تا 2.4، 4 عدد بین 2.4 تا 3.8، 4 عدد بین 3.8 تا 5.2 و ... هستند. 
</div>

In [208]:
np.convolve([1,2,3,3,2,1,-1,-2,-5,-7,-6], [0.5, 0.5]) # کانولوشن دو آرایه با هم. محاسبه میانگین متحرک 2 تایی آرایه اول

array([ 0.5,  1.5,  2.5,  3. ,  2.5,  1.5,  0. , -1.5, -3.5, -6. , -6.5,
       -3. ])

<div class="titr_fa">
10- ذخیره و بارگذاری فایل
</div>

In [226]:
np.save('fftresult.npy', u); # ذخیره ماتریس در فایل

In [227]:
v = np.load('fftresult.npy') # بارگذاری فایل در یک ماتریس جدید
u - v # آیا ماتریس ذخیره شده با ماتریس بارگذاری شده برابر است

array([0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j,
       0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j])

<div class="norm_fa">
برای ذخیره و بارگذاری فایلهای نرم افزار MATLAB از کتابخانه scipy.io و تابع savemat و loadmat میتونیم استفاده کنیم. برای خواندن فایلهای CSV (Commad Seperated Values) نامپای یک تابع به نام loadtxt داره که البته فایل فقط باید شامل عدد باشه و کاراکترهای متنی براش مشکل ایجاد میکنن. برای اینکار در مطالب بعدی کتابخونه pandas رو معرفی میکنیم که روی numpy بنا شده و عملیات ممکن در اون رو برای کار روی داده ها و خصوصا جدولهای داده گسترش میده و امکاناتی شبیه به SQL در اختیار ما قرار میده.
</div>